test/codegen: add UEFI cross-compilation test

dvdhrm opened the pull request:
Add a new codegen test which cross-compiles a UEFI Hello-World program and verifies that an entry-point is emitted.

The UEFI code-generation has broken in nightly before, see for instance:

    commit c1b2fd2121091df67d6a298f535b2db83e67dde5
    Merge: faccb0f07a5 6e77729ed5a
    Author: Dylan DPC <dylan.dpc@gmail.com>
    Date:   Tue May 5 01:49:44 2020 +0200

        Rollup merge of #71881 - IsaacWoods:master, r=petrochenkov

        Correctly handle UEFI targets as Windows-like when emitting sections for LLVM bitcode

This commit adds a very simple UEFI program and verifies that cross-compilation with the built-in `x86_64-unknown-uefi` target works. No runtime testing, nor any elaborate code-validation is done. This simply serves as base test that verifies compilation works and an entry point is emitted.

Usually, you would use `libcore`, but in this test we do not want to trigger re-compilation of `libcore` for `x86_64-unknown-uefi`. Hence, we pull in just enough of `libcore` to make the test compile.

This UEFI program is fully functional and prints `Hello World\n` to the console (UEFI standard output console). You can run it from your UEFI BootManager or from EfiShell. You might want to add `loop {}` before returning from `main()` to actually see the message (from EfiShell it works unchanged).

The imported UEFI definitions are explicitly kept simple. For more elaborate use-cases you would want the full UEFI specification (see the `r_efi` crate for a standalone import of the base UEFI specification).

rust-highfive commented:
r? @nikomatsakis

(rust_highfive has picked a reviewer for you, use r? to override)

jonas-schievink commented:
This is currently a tier 3 platform, and we usually don't even have in-tree tests for tier 2 platforms (aside from Android and some embedded Arm targets), so maybe this should be upgraded to tier 2 first? That would also allow you to remove the `#![no_core]`, which can mean more maintenance overhead (for example when lang items are changed or new ones are added).

nikomatsakis commented:
I too was wondering about the tier level. I'm going to refer this to @joshtriplett for the moment -- I ping him to get his take on the test and what level of testing we ought to be doing.

nikomatsakis commented:
I'm not necessarily opposed to adding some limited testing, though I would prefer a test that doesn't duplicate parts of the standard library, and I'm not sure if this is the right form for that testing to take.

joshtriplett commented:
The only thing this test actually tests is that the entry point gets generated. All of the additional code to print "Hello world" does not get tested, so it shouldn't be part of this test.

I think it's reasonable to have a basic codegen test to make sure we can emit an `efi_main`.

In the future, if we're going to test calls to UEFI, we should actually run the test program in a UEFI environment (like qemu with OVMF) and check its behavior. That would be part of raising a UEFI target to a higher tier.

dvdhrm commented:
> I'm not necessarily opposed to adding some limited testing, though I would prefer a test that doesn't duplicate parts of the standard library, and I'm not sure if this is the right form for that testing to take.

Just to be clear, the "duplication" part we speak of is this:

```rust
#[lang = "sized"]
pub trait Sized {}
#[lang = "freeze"]
pub trait Freeze {}
#[lang = "copy"]
pub trait Copy {}

impl<T: ?Sized> Copy for *mut T {}
```

And with a short `git grep` there are roughly ~15 other tests doing something similar (I actually copied it from one of the codegen tests). I am not saying this is an argument in favor of it, just listing it so we know what parts this would apply to.

> This is currently a tier 3 platform, and we usually don't even have in-tree tests for tier 2 platforms (aside from Android and some embedded Arm targets), so maybe this should be upgraded to tier 2 first? That would also allow you to remove the #![no_core], which can mean more maintenance overhead (for example when lang items are changed or new ones are added).

Oh, I thought I had to first add tests to suggest promoting the target to tier2. I have no objections to promoting it, though I am a bit lost on how to do that. Is it as simple as providing a PR patching `src/tools/build-manifest/src/main.rs` and then notifying the rust-forge people about it?

Secondly, you sure whether this would really help here? The **uefi** targets cannot compile `libstd`, nor `rustc`, nor any tool that you would expect. And it is very unlikely that they ever will. The platform is just too limited. Therefore, the only way to build is to cross-compile. Hence, I added this test that verifies cross-compiling works from all hosts.
If the target was a tier2 target, would the test-suite then cross-build all tests for that target? Lots of tests require libstd, how would that work?
I am a bit lost, sorry :-/

> The only thing this test actually tests is that the entry point gets generated. All of the additional code to print "Hello world" does not get tested, so it shouldn't be part of this test.

That is not entirely true. The test verifies compilation succeeds. As the commit-messages mentions, this has broken in the past, so I thought I add a test that catches such breakage so it does not happen again. Note that the test-suite already contains a cross-compile test for uefi targets (codegen/abi-efiapi.rs) which did not catch this breakage, as all it does is emit a stub symbol. This is why I added this slightly bigger test that emits a arguably short, but definitely typical UEFI code path.

I am fine with dropping the print-call. Alternatively, I could also add a codegen check that verifies a function-call is actually emitted, if that would justify keeping the code?

My intention was to have a test that verifies `rustc` can cross-compile UEFI `hello-world`. Rust has many tests to verify compilation fails, but I was not sure where I can add tests that verify compilation succeeds. Hence, I turned it into a codegen test and added the symbol-check because at least one CHECK line is required (and the check actually makes sense and is a nice addon).

> I think it's reasonable to have a basic codegen test to make sure we can emit an efi_main.
>
> In the future, if we're going to test calls to UEFI, we should actually run the test program in a UEFI environment (like qemu with OVMF) and check its behavior. That would be part of raising a UEFI target to a higher tier.

This sounds reasonable to me. The rust-forge doc lists tier2 platforms as "guaranteed to build, not necessarily a working build" (paraphrased), so not sure actual runtime tests are required, but I am all for it, so fine with me!

> Could you please drop any of the test code not strictly necessary to test that we emit efi_main?

As mentioned above, such a test already exists in form of `codegen/abi-efiapi.rs`. Arguably, it compiles to a library rather than binary, so maybe that would have already caught the bug I mentioned in the commit-message. Lets see what you think of my other comments and then re-evaluate again. I will gladly add other tests as well.

joshtriplett commented:
On Tue, May 19, 2020 at 12:20:21AM -0700, David Rheinsberg wrote:
> > I'm not necessarily opposed to adding some limited testing, though I would prefer a test that doesn't duplicate parts of the standard library, and I'm not sure if this is the right form for that testing to take.
> 
> Just to be clear, the "duplication" part we speak of is this:
> 
> ```rust
> #[lang = "sized"]
> pub trait Sized {}
> #[lang = "freeze"]
> pub trait Freeze {}
> #[lang = "copy"]
> pub trait Copy {}
> 
> impl<T: ?Sized> Copy for *mut T {}
> ```
> 
> And with a short `git grep` there are roughly ~15 other tests doing something similar (I actually copied it from one of the codegen tests). I am not saying this is an argument in favor of it, just listing it so we know what parts this would apply to.

That part seems fine to me, if it's required to compile a `no_core`
test; it's the SimpleTextOutputProtocol definitions I'd suggest not
adding.

> The **uefi** targets cannot compile `libstd`, nor `rustc`, nor any tool that you would expect. And it is very unlikely that they ever will.

I'm hoping that it'll support `std` eventually.

Supporting `rustc` seems theoretically possible, though of rather low
value. But supporting `std` seems quite useful and reasonable.

> The platform is just too limited. Therefore, the only way to build is to cross-compile. Hence, I added this test that verifies cross-compiling works from all hosts.

There are absolutely platforms that don't run tools, and people always
cross-compile for those.

> > The only thing this test actually tests is that the entry point gets generated. All of the additional code to print "Hello world" does not get tested, so it shouldn't be part of this test.
> 
> That is not entirely true. The test verifies compilation succeeds. As the commit-messages mentions, this has broken in the past, so I thought I add a test that catches such breakage so it does not happen again. Note that the test-suite already contains a cross-compile test for uefi targets (codegen/abi-efiapi.rs) which did not catch this breakage, as all it does is emit a stub symbol. This is why I added this slightly bigger test that emits a arguably short, but definitely typical UEFI code path.
> 
> I am fine with dropping the print-call. Alternatively, I could also add a codegen check that verifies a function-call is actually emitted, if that would justify keeping the code?
> 
> My intention was to have a test that verifies `rustc` can cross-compile UEFI `hello-world`. Rust has many tests to verify compilation fails, but I was not sure where I can add tests that verify compilation succeeds. Hence, I turned it into a codegen test and added the symbol-check because at least one CHECK line is required (and the check actually makes sense and is a nice addon).

Verifying compilation makes sense, but as far as I can tell, nothing in
the test verifies that the "Hello World" actually works, or that the
UEFI structures match expectations. I don't think it makes sense to have
something in the test suite that isn't actually tested.

Verifying a function call could make sense, but would make the test more
fragile.

Anything you can actually *verify* in the test is fine to keep; I just
wouldn't suggest keeping sample code that the test doesn't actually
validate.

> > I think it's reasonable to have a basic codegen test to make sure we can emit an efi_main.
> >
> > In the future, if we're going to test calls to UEFI, we should actually run the test program in a UEFI environment (like qemu with OVMF) and check its behavior. That would be part of raising a UEFI target to a higher tier.
> 
> This sounds reasonable to me. The rust-forge doc lists tier2 platforms as "guaranteed to build, not necessarily a working build" (paraphrased), so not sure actual runtime tests are required, but I am all for it, so fine with me!

Right, that would be something that would happen as part of raising UEFI
to a higher tier (which I'd love to see someday, if someone is willing
to maintain it).

nikomatsakis commented:
(Quick note from me, the duplication doesn't seem too bad, if it's necessary.)

dvdhrm commented:
I pushed a reduced version as suggested. Thanks for the feedback!

> I'm hoping that it'll support `std` eventually. Supporting `rustc` seems theoretically possible, though of rather low value. But supporting `std` seems quite useful and reasonable.

I am less convinced of the value of `std` for UEFI. It would be quite reduced, as UEFI simply does not allow for a lot of the more complex OS interfaces (e.g., it explicitly disallows interrupts, or anything but the boot-cpu). Additionally, UEFI has a capability-based model from ground up, which you would loose with the rust standard library (which requires quite a bit of global state).

Anyway, there are good arguments in both directions, so I wouldn't be surprised if someone ports it.

> Verifying compilation makes sense, but as far as I can tell, nothing in the test verifies that the "Hello World" actually works, or that the UEFI structures match expectations. I don't think it makes sense to have something in the test suite that isn't actually tested. Verifying a function call could make sense, but would make the test more fragile. Anything you can actually *verify* in the test is fine to keep; I just wouldn't suggest keeping sample code that the test doesn't actually validate.

In the rustc repository we have many tests that verify compilation fails, and those seem to have value. Why is a test that verifies compilation succeeds not considered valuable?
I don't mind dropping the bits, but I am not sure I can follow your argument. Though, I do see that the value of such a test is quite low.

> Right, that would be something that would happen as part of raising UEFI to a higher tier (which I'd love to see someday, if someone is willing to maintain it).

We here at Red Hat use rust-stable successfully on UEFI with both liballoc and libcore. We do not ship any products yet, though. But we cannot ship nightly-builds, and liballoc only recently became stable. Anyway, if there is possibility to raise the UEFI target to tier2, we would gladly maintain it. I did find it hard to get information about the process, though.

Thanks for the feedback and for the patience!

JohnCSimon commented:
ping from triage:
@dvdhrm is this ready for review?

dvdhrm commented:
> ping from triage:
> @dvdhrm is this ready for review?

Yes it is. Do I need to mark it as such somehow?

Muirrum commented:
@joshtriplett This is a triage bump.

Dylan-DPC commented:
r? @Mark-Simulacrum

Mark-Simulacrum commented:
r? @joshtriplett 

@joshtriplett is a good reviewer here, I think, and has context it seems, so I'm going to defer to them.

Dylan-DPC commented:
@joshtriplett any updates?

jyn514 commented:
Ping @joshtriplett, this is waiting on your review.

Dylan-DPC commented:
@Mark-Simulacrum anything we can do about this pr?

joshtriplett commented:
I'm sorry that this waited so long to get further feedback.

When this was originally filed, we didn't yet have the [Target Tier Policy](https://doc.rust-lang.org/nightly/rustc/target-tier-policy.html) ratified. So, much of my commentary in this thread was just my thoughts on how we should handle the target, and that didn't make for a definitive answer.

Now that that policy has been ratified, I think we do have a clear answer:

This test would require having the corresponding uefi target available. Since this is a tier 3 target, we don't block CI on it building, and this test *would* block CI on the uefi target building. So, I don't think we should apply this, *yet*.

That said, now that we have a policy in place, I think it'd be entirely appropriate to work towards the UEFI targets reaching tier 2. (That does *not* require them to implement `std`; it's entirely reasonable to have a `no_std`-only target.) I think the UEFI targets should be able to meet all of the requirements, assuming someone is willing to commit to maintaining them at tier 2. And with the targets at tier 2, we could then absolutely add this, since it's just a test to confirm that you can build for the target, rather than checking if the target passes the Rust testsuite.
